packages/vite/src/node/server/middlewares/indexHtml.ts

```diff
@@ -114,43 +114,45 @@ const processNodeUrl = (
   originalUrl?: string,
   server?: ViteDevServer,
 ): string | undefined => {
-  if (server?.moduleGraph) {
-    const mod = server.moduleGraph.urlToModuleMap.get(url)
-    if (mod && mod.lastHMRTimestamp > 0) {
-      url = injectQuery(url, `t=${mod.lastHMRTimestamp}`)
+  // prefix with base (dev only, base is never relative)
+  const replacer = (url: string) => {
+    if (server?.moduleGraph) {
+      const mod = server.moduleGraph.urlToModuleMap.get(url)
+      if (mod && mod.lastHMRTimestamp > 0) {
+        url = injectQuery(url, `t=${mod.lastHMRTimestamp}`)
+      }
     }
-  }
 
-  if (
-    (url[0] === '/' && url[1] !== '/') ||
-    // #3230 if some request url (localhost:3000/a/b) return to fallback html, the relative assets
-    // path will add `/a/` prefix, it will caused 404.
-    //
-    // skip if url contains `:` as it implies a url protocol or Windows path that we don't want to replace.
-    //
-    // rewrite `./index.js` -> `localhost:5173/a/index.js`.
-    // rewrite `../index.js` -> `localhost:5173/index.js`.
-    // rewrite `relative/index.js` -> `localhost:5173/a/relative/index.js`.
-    ((url[0] === '.' || (wordCharRE.test(url[0]) && !url.includes(':'))) &&
-      originalUrl &&
-      originalUrl !== '/' &&
-      htmlPath === '/index.html')
-  ) {
-    // prefix with base (dev only, base is never relative)
-    const replacer = (url: string) => {
+    if (
+      (url[0] === '/' && url[1] !== '/') ||
+      // #3230 if some request url (localhost:3000/a/b) return to fallback html, the relative assets
+      // path will add `/a/` prefix, it will caused 404.
+      //
+      // skip if url contains `:` as it implies a url protocol or Windows path that we don't want to replace.
+      //
+      // rewrite `./index.js` -> `localhost:5173/a/index.js`.
+      // rewrite `../index.js` -> `localhost:5173/index.js`.
+      // rewrite `relative/index.js` -> `localhost:5173/a/relative/index.js`.
+      ((url[0] === '.' || (wordCharRE.test(url[0]) && !url.includes(':'))) &&
+        originalUrl &&
+        originalUrl !== '/' &&
+        htmlPath === '/index.html')
+    ) {
       const devBase = config.base
       const fullUrl = path.posix.join(devBase, url)
       if (server && shouldPreTransform(url, config)) {
         preTransformRequest(server, fullUrl, devBase)
       }
       return fullUrl
+    } else {
+      return url
     }
```

Comment:
Perhaps we should return `undefined` here, or return `undefined` from the `processNodeUrl` function so that if the URL didn't change, we can skip calling magic-string overwrite:
https://github.com/vitejs/vite/blob/ac2503d52cfc071b5be07485183708dee23b82c5/packages/vite/src/node/server/middlewares/indexHtml.ts#L241-L251

Reply:
Also seems like we need to handle `injectQuery` above to update the URLs if it changed

Reply:
Done here https://github.com/vitejs/vite/pull/14870/commits/68e09595026d92990aebbaa175eebe34afeda53e, I think the logic is simpler if we avoid returning `undefined`.